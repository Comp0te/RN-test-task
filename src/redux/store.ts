import { createStore, applyMiddleware, combineReducers } from 'redux';
import { combineEpics, createEpicMiddleware } from 'redux-observable';
import { composeWithDevTools } from 'redux-devtools-extension';

import { requestsReducer } from './requests/reducers';

import { requestsEpics } from './requests/epics';

export const rootReducer = combineReducers({
  requests: requestsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

const rootEpic = combineEpics(
  ...requestsEpics,
);

const epicMiddleware = createEpicMiddleware();

const middleware = [
  epicMiddleware,
];

const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(...middleware)),
);

epicMiddleware.run(rootEpic);

export default store;
