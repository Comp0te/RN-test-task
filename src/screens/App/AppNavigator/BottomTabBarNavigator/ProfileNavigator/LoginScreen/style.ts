import { StyleSheet } from 'react-native';
import { mainColor, greyColor } from '../../../../../../app.style';

const style = StyleSheet.create({
  safeArea: {
    flex: 1,
  },
  root: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: 95,
    paddingBottom: 20,
    paddingLeft: 40,
    paddingRight: 40,
    minHeight: 460,
  },
  fieldWrapper: {
    width: '100%',
  },
  passwordWrapper: {
    width: '100%',
    marginBottom: 16,
  },
  signInWrapper: {
    width: '100%',
    height: 40,
    marginBottom: 10,
  },
  singUpWrapper: {
    position: 'absolute',
    bottom: 10,
    flexDirection: 'row',
    alignItems: 'center',
  },
  image: {
    width: 160,
    height: 33,
  },
  text: {
    fontWeight: '400',
    fontSize: 11,
    color: greyColor,
  },
  textLink: {
    color: mainColor,
  },
});

export default style;