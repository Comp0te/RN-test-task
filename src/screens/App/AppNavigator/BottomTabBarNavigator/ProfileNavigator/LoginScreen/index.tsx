import React, { useCallback, useEffect } from 'react';
import { compose, Dispatch } from 'redux';
import { connect } from 'react-redux';
import style from './style';

import {
  View, Text, TouchableOpacity, SafeAreaView,
  KeyboardAvoidingView, ScrollView,
} from 'react-native';
import { Field, InjectedFormProps, reduxForm } from 'redux-form';
import FieldInput from '../../../../../../shared/components/FieldInput';
import CommonButton from '../../../../../../shared/components/CommonButton';

import required from '../../../../../../shared/validators/required';
import { passwordLogin } from '../../../../../../shared/validators/password';
import { userNameLetters } from '../../../../../../shared/validators/userName';
import { maxLengthUserName, minLengthUserName } from '../../../../../../shared/validators/lenght';

import { AuthInput } from '../../../../../../shared/services/auth.service';
import { RootState } from '../../../../../../redux/store';
import { NavigationInjectedProps } from 'react-navigation';

import navService from '../../../../../../../src/shared/services/nav.service';
import * as loginAC from '../../../../../../redux/requests/requestsEntities/auth/login/AC';
import { getIsLoginRequestLoading } from '../../../../../../redux/requests/selectors';
import { getIsAuthUser } from '../../../../../../redux/auth/selectors';

type LoginFormData = AuthInput;

interface StateProps {
  isLoading: boolean;
  isAuthUser: boolean;
}

const mapStateToProps = (state: RootState): StateProps => ({
  isLoading: getIsLoginRequestLoading(state),
  isAuthUser: getIsAuthUser(state),
});

type Props = StateProps & InjectedFormProps<LoginFormData> & NavigationInjectedProps;

const LoginScreen: React.FC<Props> = (props) => {
  const {handleSubmit, isLoading, destroy, navigation, isAuthUser} = props;

  useEffect(() => {
    if (isAuthUser) {
      navService.navigate('ProfileScreen');
    }
  }, []);

  const onBlur = useCallback(() => {
    destroy();
  }, []);

  useEffect(() => {
    const focusListener = navigation.addListener('willBlur', onBlur);

    return () => focusListener.remove();
  }, [onBlur]);

  const submitLogin = (values: LoginFormData, dispatch: Dispatch<loginAC.Actions>) => {
    dispatch(loginAC.Actions.login(values));
  };

  const toRegistrationScreen = () => {
    navService.navigate('RegistrationScreen');
  };

  return (
    <SafeAreaView style={style.safeArea}>
      <ScrollView contentContainerStyle={style.safeArea}>
        <KeyboardAvoidingView
          behavior='padding'
          style={style.root}
        >
          <View style={style.fieldWrapper}>
            <Field
              name='username'
              component={FieldInput}
              keyboard='default'
              placeholder='User name'
              leftIconName='user'
              maxLength={20}
              validate={[
                required,
                userNameLetters,
                maxLengthUserName,
                minLengthUserName,
              ]}
            />
          </View>
          <View style={style.passwordWrapper}>
            <Field
              name='password'
              component={FieldInput}
              placeholder='Password'
              leftIconName='lock'
              validate={[required, passwordLogin]}
              secureTextEntry={true}
            />
          </View>
          <View style={style.signInWrapper}>
            <CommonButton
              title='SIGN IN'
              loading={isLoading}
              disabled={isLoading}
              onPress={handleSubmit(submitLogin)}
              iconName='sign-in'
            />
          </View>
          <View style={style.singUpWrapper}>
            <Text style={style.text}>
              {'Do not have an account? '.toUpperCase()}
            </Text>
            <TouchableOpacity
              onPress={toRegistrationScreen}
              hitSlop={{
                top: 10,
                bottom: 10,
                right: 0,
                left: 0,
              }}
            >
              <Text style={[style.text, style.textLink]}>SIGN UP</Text>
            </TouchableOpacity>
          </View>
        </KeyboardAvoidingView>
      </ScrollView>
    </SafeAreaView>
  );
};

export default compose<React.ComponentType>(
  reduxForm<LoginFormData>({
    form: 'login',
  }),
  connect(mapStateToProps),
)(LoginScreen);
