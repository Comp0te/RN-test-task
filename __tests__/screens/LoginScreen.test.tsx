import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';
import LoginScreen from '../../src/screens/App/AppNavigator/AuthNavigator/LoginScreen';
import { Provider } from 'react-redux';
import store from '../../src/redux/store';

it('LoginScreen renders correctly', () => {
  expect(renderer.create(
    <Provider store={store}>
      <LoginScreen/>
    </Provider>,
  )).toMatchSnapshot();
});