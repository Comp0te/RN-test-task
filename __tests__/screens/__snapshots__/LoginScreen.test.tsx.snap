// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "minHeight": 460,
              "paddingBottom": 20,
              "paddingLeft": 40,
              "paddingRight": 40,
              "paddingTop": 95,
            },
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "justifyContent": "flex-start",
                "minHeight": 74,
                "paddingBottom": 5,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "paddingHorizontal": 10,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomWidth": 1,
                    "borderColor": "#86939e",
                    "flexDirection": "row",
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "height": 40,
                      "justifyContent": "center",
                      "marginLeft": 15,
                      "width": 30,
                    }
                  }
                >
                  <View>
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "justifyContent": "center",
                        }
                      }
                      theme={
                        Object {
                          "colors": Object {
                            "disabled": "hsl(208, 8%, 90%)",
                            "divider": "#bcbbc1",
                            "error": "#ff190c",
                            "grey0": "#393e42",
                            "grey1": "#43484d",
                            "grey2": "#5e6977",
                            "grey3": "#86939e",
                            "grey4": "#bdc6cf",
                            "grey5": "#e1e8ee",
                            "greyOutline": "#bbb",
                            "platform": Object {
                              "android": Object {
                                "error": "#f44336",
                                "primary": "#2196f3",
                                "secondary": "#9C27B0",
                                "success": "#4caf50",
                                "warning": "#ffeb3b",
                              },
                              "ios": Object {
                                "error": "#ff3b30",
                                "primary": "#007aff",
                                "secondary": "#5856d6",
                                "success": "#4cd964",
                                "warning": "#ffcc00",
                              },
                            },
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#8F0CE8",
                            "success": "#52c41a",
                            "warning": "#faad14",
                          },
                        }
                      }
                      underlayColor="white"
                    >
                      <Text
                        allowFontScaling={false}
                        style={
                          Array [
                            Object {
                              "color": "#BEBEBE",
                              "fontSize": 20,
                            },
                            Object {
                              "backgroundColor": "transparent",
                            },
                            Object {
                              "fontFamily": "FontAwesome",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            Object {},
                          ]
                        }
                        testID="iconIcon"
                      >
                        
                      </Text>
                    </View>
                  </View>
                </View>
                <TextInput
                  allowFontScaling={true}
                  keyboard="default"
                  maxLength={20}
                  name="username"
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onDragStart={[Function]}
                  onDrop={[Function]}
                  onFocus={[Function]}
                  placeholder="User name"
                  placeholderTextColor="#BEBEBE"
                  rejectResponderTermination={true}
                  style={
                    Object {
                      "alignSelf": "center",
                      "borderWidth": 0,
                      "color": "#860483",
                      "flex": 1,
                      "fontSize": 14,
                      "fontWeight": "400",
                      "height": 40,
                      "minHeight": 40,
                      "paddingRight": 25,
                      "width": "100%",
                    }
                  }
                  testID="RNE__Input__text-input"
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 16,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "justifyContent": "flex-start",
                "minHeight": 74,
                "paddingBottom": 5,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "paddingHorizontal": 10,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomWidth": 1,
                    "borderColor": "#86939e",
                    "flexDirection": "row",
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "height": 40,
                      "justifyContent": "center",
                      "marginLeft": 15,
                      "width": 30,
                    }
                  }
                >
                  <View>
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "justifyContent": "center",
                        }
                      }
                      theme={
                        Object {
                          "colors": Object {
                            "disabled": "hsl(208, 8%, 90%)",
                            "divider": "#bcbbc1",
                            "error": "#ff190c",
                            "grey0": "#393e42",
                            "grey1": "#43484d",
                            "grey2": "#5e6977",
                            "grey3": "#86939e",
                            "grey4": "#bdc6cf",
                            "grey5": "#e1e8ee",
                            "greyOutline": "#bbb",
                            "platform": Object {
                              "android": Object {
                                "error": "#f44336",
                                "primary": "#2196f3",
                                "secondary": "#9C27B0",
                                "success": "#4caf50",
                                "warning": "#ffeb3b",
                              },
                              "ios": Object {
                                "error": "#ff3b30",
                                "primary": "#007aff",
                                "secondary": "#5856d6",
                                "success": "#4cd964",
                                "warning": "#ffcc00",
                              },
                            },
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#8F0CE8",
                            "success": "#52c41a",
                            "warning": "#faad14",
                          },
                        }
                      }
                      underlayColor="white"
                    >
                      <Text
                        allowFontScaling={false}
                        style={
                          Array [
                            Object {
                              "color": "#BEBEBE",
                              "fontSize": 20,
                            },
                            Object {
                              "backgroundColor": "transparent",
                            },
                            Object {
                              "fontFamily": "FontAwesome",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            Object {},
                          ]
                        }
                        testID="iconIcon"
                      >
                        
                      </Text>
                    </View>
                  </View>
                </View>
                <TextInput
                  allowFontScaling={true}
                  name="password"
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onDragStart={[Function]}
                  onDrop={[Function]}
                  onFocus={[Function]}
                  placeholder="Password"
                  placeholderTextColor="#BEBEBE"
                  rejectResponderTermination={true}
                  secureTextEntry={true}
                  style={
                    Object {
                      "alignSelf": "center",
                      "borderWidth": 0,
                      "color": "#860483",
                      "flex": 1,
                      "fontSize": 14,
                      "fontWeight": "400",
                      "height": 40,
                      "minHeight": 40,
                      "paddingRight": 25,
                      "width": "100%",
                    }
                  }
                  testID="RNE__Input__text-input"
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "height": 40,
              "marginBottom": 30,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "borderRadius": 3,
              }
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#860483",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "marginRight": 10,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                    theme={
                      Object {
                        "colors": Object {
                          "disabled": "hsl(208, 8%, 90%)",
                          "divider": "#bcbbc1",
                          "error": "#ff190c",
                          "grey0": "#393e42",
                          "grey1": "#43484d",
                          "grey2": "#5e6977",
                          "grey3": "#86939e",
                          "grey4": "#bdc6cf",
                          "grey5": "#e1e8ee",
                          "greyOutline": "#bbb",
                          "platform": Object {
                            "android": Object {
                              "error": "#f44336",
                              "primary": "#2196f3",
                              "secondary": "#9C27B0",
                              "success": "#4caf50",
                              "warning": "#ffeb3b",
                            },
                            "ios": Object {
                              "error": "#ff3b30",
                              "primary": "#007aff",
                              "secondary": "#5856d6",
                              "success": "#4cd964",
                              "warning": "#ffcc00",
                            },
                          },
                          "primary": "#2089dc",
                          "searchBg": "#303337",
                          "secondary": "#8F0CE8",
                          "success": "#52c41a",
                          "warning": "#faad14",
                        },
                      }
                    }
                    underlayColor="white"
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": "#FFFFFF",
                            "fontSize": 20,
                          },
                          Object {
                            "backgroundColor": "transparent",
                          },
                          Object {
                            "fontFamily": "FontAwesome",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                      testID="iconIcon"
                    >
                      
                    </Text>
                  </View>
                </View>
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 18,
                      "paddingBottom": 1,
                      "paddingTop": 2,
                      "textAlign": "center",
                    }
                  }
                >
                  SIGN IN
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 0,
              "right": 0,
              "top": 10,
            }
          }
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#860483",
                "fontSize": 12,
                "fontWeight": "400",
              }
            }
          >
            SIGN IN LATER
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "bottom": 20,
              "flexDirection": "row",
              "position": "absolute",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#BEBEBE",
                "fontSize": 11,
                "fontWeight": "400",
              }
            }
          >
            DO NOT HAVE AN ACCOUNT? 
          </Text>
          <View
            accessible={true}
            hitSlop={
              Object {
                "bottom": 10,
                "left": 0,
                "right": 0,
                "top": 10,
              }
            }
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#860483",
                  "fontSize": 12,
                  "fontWeight": "400",
                }
              }
            >
              SIGN UP
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
